File: insurance.R
# load packages
install.packages("readr")
library(readr)
install.packages("CatEncoders")
library(CatEncoders)
install.packages("fastDummies")
library(fastDummies)

# load data
insuranceData <- read_csv("C:/Dataset/Cluster/AutoInsurance.csv")
insDataCopy <- insuranceData

# checking for null data
sum(is.na(insDataCopy))

# label encoding for -  ("Coverage","Education","Effective To Date","EmploymentStatus","Vehicle Class","Vehicle Size")
lblCoverage <- LabelEncoder.fit(insuranceData$Coverage)
insDataCopy$Coverage<- transform(lblCoverage,insuranceData$Coverage)
# Education encoding
lblEducation <- LabelEncoder.fit(insuranceData$Education)
insDataCopy$Education<- transform(lblEducation,insuranceData$Education)
# Effective To Date encoding
lblDate <- LabelEncoder.fit(insuranceData$`Effective To Date`)
insDataCopy$`Effective To Date`<- transform(lblDate,insuranceData$`Effective To Date`)
# EmploymentStatus encoding
lblEmploymentStatus <- LabelEncoder.fit(insuranceData$EmploymentStatus)
insDataCopy$EmploymentStatus<- transform(lblEmploymentStatus,insuranceData$EmploymentStatus)
# Vehicle Class encoding
lblClass <- LabelEncoder.fit(insuranceData$`Vehicle Class`)
insDataCopy$`Vehicle Class`<- transform(lblClass,insuranceData$`Vehicle Class`)
# Vehicle Size encoding
lblSize <- LabelEncoder.fit(insuranceData$`Vehicle Size`)
insDataCopy$`Vehicle Size`<- transform(lblSize,insuranceData$`Vehicle Size`)



#one hot encoding
processedInsData <- dummy_cols(insDataCopy, 
                                 select_columns = c("State","Response","Gender","Location Code","Marital Status","Policy","Policy Type","Renew Offer Type","Sales Channel"), 
                                 remove_first_dummy =  TRUE, 
                                 remove_most_frequent_dummy = FALSE, 
                                 remove_selected_columns = TRUE)

summary(processedInsData)
# take 5000 rows only
iData <- processedInsData[5000,]
# distance matrix
distInsData <- dist(processedInsData, method = "euclidean")

clusterData <- hclust(distInsData, method = "complete")

# plot the data - Dendrogram
plot(clusterData, hang = -1)

groupData <- cutree(clusterData, k = 4)

rect.hclust(clusterData, k = 4, border = 'red')

# creating final data
data <- as.matrix(groupData)
finalData <- data.frame(data, insuranceData)

# write to file
write.csv(finalData, "insurance.csv")

aggregate(finalData, list(finalData$data), mean)
# so group 3 gives maximum premium while group 1 gives minimum premium

File: insurance.py
# load packages
import pandas as pd
import scipy.cluster.hierarchy as sch
from sklearn.cluster import AgglomerativeClustering
from sklearn.preprocessing import LabelEncoder

# load data
insuranceData = pd.read_csv(r"C:\Dataset\Cluster\AutoInsurance.csv")
insData = insuranceData

# checking for na value
insuranceData.isna()
insuranceData.info()

# label encoding for -  ("Coverage","Education","Effective To Date","EmploymentStatus","Vehicle Class","Vehicle Size")
encoder = LabelEncoder()

insuranceData['Coverage']= encoder.fit_transform(insuranceData['Coverage'])
insuranceData['Education']= encoder.fit_transform(insuranceData['Education'])
insuranceData['Effective To Date']= encoder.fit_transform(insuranceData['Effective To Date'])
insuranceData['EmploymentStatus']= encoder.fit_transform(insuranceData['EmploymentStatus'])
insuranceData['Vehicle Class']= encoder.fit_transform(insuranceData['Vehicle Class'])
insuranceData['Vehicle Size']= encoder.fit_transform(insuranceData['Vehicle Size'])

# get dummy var data
listOfCol = ["State","Response","Gender","Location Code","Marital Status","Policy","Policy Type","Renew Offer Type","Sales Channel"]
dummyInsuranceData = pd.get_dummies(data= insuranceData,columns = listOfCol)

dummyInsuranceData.describe()

# standardize data
def colVariableStandardization(colVariable):
    return (colVariable - colVariable.mean()/colVariable.std())
stdData = colVariableStandardization(dummyInsuranceData.iloc[:, 1:])

# distance matrix - pairwise distance, dendrogram
gData = sch.linkage(stdData, method = 'complete', metric = 'euclidean')

sch.dendrogram(gData)

# agglomerative clustering

data = AgglomerativeClustering(n_clusters= 4, affinity='euclidean', linkage='complete').fit(gData)
data.labels_

clusterLabels = pd.Series(data.labels_)
clusterLabels

insData["insuranceCluster"] = clusterLabels

cols = insData.columns.tolist()
cols
cols = cols[-1:] + cols[:-1]
insData = insData[cols]

insData.to_csv("insuranceData.csv", encoding = "utf-8")

